File: src/shared/projects/project-file.ts
/**
 * The project file: a studio project's PORTABLE half on disk, as JSON.
 *
 * It is the same split the document model already makes (project-types.ts):
 * everything reusable with any media travels — composition settings, overlay
 * elements, guides, the grade stack (custom `.cube` text inlined, so a shared
 * look grades identically), the title theme, the export matrix. Nothing bound
 * to this machine does: no directory handle, no media list, no thumbnail, no
 * document id. A project file is a template you can mail, not a backup of a
 * folder.
 *
 * `version` is the ProjectDoc version the file was written at, so an older
 * file replays the same migration chain a stored document does. A file from a
 * NEWER version is refused rather than half-read — silently dropping fields we
 * do not understand would quietly damage someone's project.
 *
 * Pure and DOM-free: parsing, validating and applying are testable; picking
 * and downloading files stay in the UI.
 */

import type { OverlayElement } from '../overlay/overlay-types';
import type { GuidesState } from '../overlay/guides';
import { DEFAULT_GUIDES } from '../overlay/guides';
import type { StyleTheme } from '../overlay/title-styles';
import type { Scene } from '../overlay/scenes';
import type { SavedLutLayer } from '../lut/use-lut-stack';
import type { OutputTransform } from '../lut/transfer';
import { NO_SHIFT } from '../telemetry/time-format';
import { defaultVariants } from './export-variants';
import {
  PROJECT_DOC_VERSION,
  migrateProjectDoc,
  type ExportPrefs,
  type ProjectDoc,
  type ProjectSettings,
} from './project-types';

/** Marks the file as ours; a stray `.json` is rejected on it. */
export const PROJECT_FILE_KIND = 'atelier/studio-project';

/** Double extension: recognisable at a glance, still a plain `.json`. */
export const PROJECT_FILE_EXTENSION = '.atelier.json';

/** What the file picker accepts. */
export const PROJECT_FILE_ACCEPT = '.json,application/json';

/** The portable half of a project — what a project file carries. */
export interface ProjectPortable {
  settings: ProjectSettings;
  elements: OverlayElement[];
  guides: GuidesState;
  lutStack: SavedLutLayer[];
  outputTransform: OutputTransform;
  theme: StyleTheme | null;
  scenes: Scene[];
  exportPrefs: ExportPrefs;
}

export interface ProjectFile extends ProjectPortable {
  kind: typeof PROJECT_FILE_KIND;
  /** ProjectDoc version this was written at — drives the migration on read. */
  version: number;
  /** The project's name when it was exported; informative, never binding. */
  name: string;
  /** ISO timestamp, for the human reading the file. */
  exportedAt: string;
}

export type ParseResult =
  | { ok: true; file: ProjectFile }
  | { ok: false; error: string };

/** Build the file from a project document (or from the editor's live state). */
export function toProjectFile(
  source: ProjectPortable & { name: string },
  exportedAt: number = Date.now(),
): ProjectFile {
  // The cadence correction stays home: it is measured against ONE clip's
  // telemetry (and carries that clip's id), so travelling with a template it
  // could only misapply — a 4× override silently quadrupling another clip's
  // speeds. `parseProjectFile` drops it on the way in; not writing it keeps the
  // file honest about what it can restore.
  const portableSettings = structuredClone(source.settings);
  delete portableSettings.timeScale;
  return {
    kind: PROJECT_FILE_KIND,
    version: PROJECT_DOC_VERSION,
    name: source.name,
    exportedAt: new Date(exportedAt).toISOString(),
    settings: portableSettings,
    elements: structuredClone(source.elements),
    guides: structuredClone(source.guides),
    lutStack: structuredClone(source.lutStack),
    outputTransform: source.outputTransform,
    theme: structuredClone(source.theme),
    scenes: structuredClone(source.scenes ?? []),
    exportPrefs: structuredClone(source.exportPrefs),
  };
}

/** Indented on purpose: the file is meant to be readable and diffable. */
export function serializeProjectFile(file: ProjectFile): string {
  return `${JSON.stringify(file, null, 2)}\n`;
}

/** `Vol du soir` → `vol-du-soir.atelier.json`. */
export function projectFileName(name: string): string {
  const slug = name
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 60);
  return `${slug || 'project'}${PROJECT_FILE_EXTENSION}`;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

/**
 * Read a file's text into a project file, or explain why it isn't one. The
 * input comes from the user's disk and may be anything at all: this never
 * throws, and every rejection says something a human can act on.
 */
export function parseProjectFile(text: string): ParseResult {
  let raw: unknown;
  try {
    raw = JSON.parse(text);
  } catch {
    return { ok: false, error: 'That file is not valid JSON.' };
  }
  if (!isRecord(raw)) {
    return { ok: false, error: 'That file is not an Atelier project file.' };
  }
  if (raw.kind !== PROJECT_FILE_KIND) {
    return {
      ok: false,
      error: 'That file is not an Atelier project file (wrong or missing kind).',
    };
  }
  const version = typeof raw.version === 'number' ? raw.version : 0;
  if (version > PROJECT_DOC_VERSION) {
    return {
      ok: false,
      error:
        `This file was written by a newer version of Atelier (format ${version}, ` +
        `this one reads up to ${PROJECT_DOC_VERSION}). Update the app first.`,
    };
  }
  if (!Array.isArray(raw.elements)) {
    return { ok: false, error: 'The file has no overlay elements list.' };
  }
  if (!isRecord(raw.settings) || typeof raw.settings.aspectId !== 'string') {
    return { ok: false, error: 'The file has no project format.' };
  }

  // Shape is sound; anything optional that a past version did not write is
  // filled with the same default a migrated document gets.
  const portable: ProjectPortable = {
    settings: {
      aspectId: raw.settings.aspectId,
      timeShift: isRecord(raw.settings.timeShift)
        ? {
            minutes: Number(raw.settings.timeShift.minutes) || 0,
            days: Number(raw.settings.timeShift.days) || 0,
          }
        : { ...NO_SHIFT },
    },
    elements: raw.elements as OverlayElement[],
    guides: isRecord(raw.guides)
      ? (raw.guides as unknown as GuidesState)
      : structuredClone(DEFAULT_GUIDES),
    lutStack: Array.isArray(raw.lutStack) ? (raw.lutStack as SavedLutLayer[]) : [],
    outputTransform: (raw.outputTransform as OutputTransform) ?? 'none',
    theme: isRecord(raw.theme) ? (raw.theme as unknown as StyleTheme) : null,
    scenes: Array.isArray(raw.scenes) ? (raw.scenes as Scene[]) : [],
    exportPrefs:
      isRecord(raw.exportPrefs) && Array.isArray(raw.exportPrefs.variants)
        ? {
            fileName:
              typeof raw.exportPrefs.fileName === 'string'
                ? raw.exportPrefs.fileName
                : null,
            variants: raw.exportPrefs.variants as ExportPrefs['variants'],
          }
        : { fileName: null, variants: defaultVariants() },
  };

  // Replay the document migrations: an older file must land on the current
  // shape exactly as an older stored project does.
  const migrated = migrateProjectDoc({
    ...(portable as unknown as ProjectDoc),
    version,
  } as ProjectDoc);

  return {
    ok: true,
    file: {
      kind: PROJECT_FILE_KIND,
      version: PROJECT_DOC_VERSION,
      name: typeof raw.name === 'string' ? raw.name : '',
      exportedAt: typeof raw.exportedAt === 'string' ? raw.exportedAt : '',
      settings: migrated.settings,
      elements: migrated.elements,
      guides: migrated.guides,
      lutStack: migrated.lutStack,
      outputTransform: migrated.outputTransform,
      theme: migrated.theme,
      scenes: migrated.scenes,
      exportPrefs: migrated.exportPrefs,
    },
  };
}

/**
 * Replace a document's portable half with the file's. The bound half — id,
 * media folder, thumbnail, creation date — and the project's own NAME are
 * kept: importing settings into "Vol du soir" must not rename it after
 * whoever exported the file.
 */
export function applyProjectFile(
  doc: ProjectDoc,
  file: ProjectFile,
  now: number = Date.now(),
): ProjectDoc {
  return {
    ...doc,
    updatedAt: now,
    settings: structuredClone(file.settings),
    elements: structuredClone(file.elements),
    guides: structuredClone(file.guides),
    lutStack: structuredClone(file.lutStack),
    outputTransform: file.outputTransform,
    theme: structuredClone(file.theme),
    exportPrefs: structuredClone(file.exportPrefs),
  };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Projects/ProjectFile.swift
// The project file — a Studio project's PORTABLE half on disk as JSON,
// `<name>.atelier.json`. Port of `src/shared/projects/project-file.ts`, pure:
// building, serialising, parsing and applying; picking and saving the file
// are the app's.
//
// It is the same split the document makes (`ProjectTypes.swift`): everything
// reusable with any media travels — the format and the capture-time shift,
// the overlay elements, the guides, the grade stack (an uploaded `.cube`
// inlined, so a shared look grades identically), the film texture, the title
// theme, the intro scenes, the outro card, the export matrix. Nothing bound
// to one machine or one footage does: no folder, no media list, no trims, no
// develops, no thumbnail, no document id, no source — and not the cadence
// correction, which is measured against ONE clip and would misapply to
// another. A project file is a template you can mail, not a backup.
//
// THE FOUR PLACES (`studio.md`): anything added to the portable half must
// land in `ProjectPortable`, `toProjectFile`, `parseProjectFile` and
// `applyProjectFile` — the last was forgotten once on the web and the gallery
// import silently dropped intros. Here the four are this file's four
// functions, over the one `ProjectPortable` record.
//
// `version` is the ProjectDoc version the file was written at, so an older
// file replays the same migration chain a stored document does. A file from a
// NEWER version is refused rather than half-read. The text comes off a
// stranger's disk: parsing never throws, and every refusal is a sentence a
// person can act on.

import Foundation

/// Marks the file as ours; a stray `.json` is rejected on it.
public let projectFileKind = "atelier/studio-project"

/// Double extension: recognisable at a glance, still a plain `.json`.
public let projectFileExtension = ".atelier.json"

/// What the web's file picker accepts — the app's picker takes `.json` files.
public let projectFileAccept = ".json,application/json"

/// The portable half of a project — what a project file carries.
public struct ProjectPortable: Equatable, Sendable {
    public var settings: ProjectSettings
    public var elements: [OverlayElement]
    /// Stored elements this build cannot read, carried in place (`UnreadElement`).
    public var unreadElements: [UnreadElement]
    public var guides: GuidesState
    public var lutStack: [SavedLutLayer]
    public var outputTransform: OutputTransform
    public var lutFilm: FilmTexture?
    public var theme: StyleTheme?
    public var scenes: [OverlayScene]
    public var outro: OutroCard?
    public var exportPrefs: ExportPrefs

    public init(settings: ProjectSettings, elements: [OverlayElement], unreadElements: [UnreadElement] = [],
                guides: GuidesState, lutStack: [SavedLutLayer], outputTransform: OutputTransform,
                lutFilm: FilmTexture?, theme: StyleTheme?, scenes: [OverlayScene], outro: OutroCard?,
                exportPrefs: ExportPrefs) {
        self.settings = settings; self.elements = elements; self.unreadElements = unreadElements
        self.guides = guides; self.lutStack = lutStack; self.outputTransform = outputTransform
        self.lutFilm = lutFilm; self.theme = theme; self.scenes = scenes; self.outro = outro
        self.exportPrefs = exportPrefs
    }
}

extension ProjectDoc {
    /// The document's portable half (place one of four).
    public var portable: ProjectPortable {
        ProjectPortable(settings: settings, elements: elements, unreadElements: unreadElements, guides: guides,
                        lutStack: lutStack, outputTransform: outputTransform, lutFilm: lutFilm, theme: theme,
                        scenes: scenes, outro: outro, exportPrefs: exportPrefs)
    }
}

/// A project file: the portable half, plus what says whose and when.
public struct ProjectFile: Equatable, Sendable {
    /// Always `projectFileKind`.
    public var kind: String
    /// ProjectDoc version this was written at — drives the migration on read.
    public var version: Int
    /// The project's name when it was exported; informative, never binding.
    public var name: String
    /// ISO timestamp, for the human reading the file.
    public var exportedAt: String
    public var settings: ProjectSettings
    public var elements: [OverlayElement]
    public var unreadElements: [UnreadElement]
    public var guides: GuidesState
    public var lutStack: [SavedLutLayer]
    public var outputTransform: OutputTransform
    public var lutFilm: FilmTexture?
    public var theme: StyleTheme?
    public var scenes: [OverlayScene]
    public var outro: OutroCard?
    public var exportPrefs: ExportPrefs

    public init(version: Int = projectDocVersion, name: String, exportedAt: String, portable p: ProjectPortable) {
        kind = projectFileKind
        self.version = version; self.name = name; self.exportedAt = exportedAt
        settings = p.settings; elements = p.elements; unreadElements = p.unreadElements; guides = p.guides
        lutStack = p.lutStack; outputTransform = p.outputTransform; lutFilm = p.lutFilm; theme = p.theme
        scenes = p.scenes; outro = p.outro; exportPrefs = p.exportPrefs
    }

    /// The file's portable half.
    public var portable: ProjectPortable {
        ProjectPortable(settings: settings, elements: elements, unreadElements: unreadElements, guides: guides,
                        lutStack: lutStack, outputTransform: outputTransform, lutFilm: lutFilm, theme: theme,
                        scenes: scenes, outro: outro, exportPrefs: exportPrefs)
    }
}

/// Why a text is not a project file — a sentence a person can act on.
public struct ProjectFileError: Error, Equatable, Sendable {
    public var message: String
    public init(_ message: String) { self.message = message }
}

/// `new Date(ms).toISOString()`.
private func projectFileIsoString(_ ms: Double) -> String {
    let iso = ISO8601DateFormatter()
    iso.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return iso.string(from: Date(timeIntervalSince1970: ms / 1000))
}

// MARK: - place two: building

/// Build the file from a portable half (the editor's live state, or a
/// document's) and the project's name (place two of four).
public func toProjectFile(_ source: ProjectPortable, name: String, exportedAt: Double = nowMillis()) -> ProjectFile {
    // The cadence correction stays home: measured against ONE clip's
    // telemetry (and carrying that clip's id), travelling with a template it
    // could only misapply. `parseProjectFile` drops it on the way in too.
    var portable = source
    portable.settings.timeScale = nil
    return ProjectFile(version: projectDocVersion, name: name, exportedAt: projectFileIsoString(exportedAt),
                       portable: portable)
}

/// The file of a whole document.
public func toProjectFile(_ doc: ProjectDoc, exportedAt: Double = nowMillis()) -> ProjectFile {
    toProjectFile(doc.portable, name: doc.name, exportedAt: exportedAt)
}

extension ProjectFile {
    /// The file as JSON — the portable half and the four header fields, never
    /// a bound one (no `sourceId`, no `media`, no `thumbnail`, no `id`).
    public var json: JSONValue {
        .object([
            "kind": .string(kind),
            "version": .number(Double(version)),
            "name": .string(name),
            "exportedAt": .string(exportedAt),
            "settings": settings.json,
            "elements": projectElementsJSON(elements, unreadElements),
            "guides": guides.json,
            "lutStack": .array(lutStack.map(\.json)),
            "outputTransform": .string(outputTransform.rawValue),
            "lutFilm": lutFilm?.json ?? .null,
            "theme": theme?.json ?? .null,
            "scenes": .array(scenes.map(\.json)),
            "outro": outro?.json ?? .null,
            "exportPrefs": exportPrefs.json,
        ])
    }
}

/// Indented on purpose: the file is meant to be readable and diffable.
public func serializeProjectFile(_ file: ProjectFile) -> String {
    file.json.serialized(pretty: true) + "\n"
}

/// `Vol du soir` → `vol-du-soir.atelier.json`: accents stripped (NFD, the
/// combining marks dropped), lowercased, every run of anything but `a-z0-9`
/// one dash, the ends trimmed of dashes, then the first 60 characters.
public func projectFileName(_ name: String) -> String {
    var folded = String.UnicodeScalarView()
    for scalar in name.decomposedStringWithCanonicalMapping.unicodeScalars
    where !(0x300...0x36F).contains(scalar.value) {
        folded.append(scalar)
    }
    var slug = ""
    var pendingDash = false
    for scalar in String(folded).lowercased().unicodeScalars {
        let v = scalar.value
        if (0x61...0x7A).contains(v) || (0x30...0x39).contains(v) {
            if pendingDash { slug.append("-") }
            pendingDash = false
            slug.unicodeScalars.append(scalar)
        } else if !slug.isEmpty {
            // A leading run is trimmed; a trailing one is never written.
            pendingDash = true
        }
    }
    slug = String(slug.prefix(60))
    return "\(slug.isEmpty ? "project" : slug)\(projectFileExtension)"
}

// MARK: - place three: parsing

/// JavaScript's `Number(x) || 0` over a JSON value: a number, a numeric
/// string, a boolean; anything else (and NaN) is 0.
private func projectFileNumber(_ v: JSONValue?) -> Double {
    guard let v else { return 0 }
    switch v {
    case .number(let n): return n.isFinite ? n : 0
    case .bool(let b): return b ? 1 : 0
    case .string(let s):
        let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.isEmpty { return 0 }
        guard let n = Double(t), n.isFinite else { return 0 }
        return n
    default: return 0
    }
}

/// JavaScript's `${n}`: an integer-valued number prints without `.0`.
private func projectFileNumberText(_ x: Double) -> String {
    if x.isFinite, x == x.rounded(), abs(x) < 1e15 { return String(Int64(x)) }
    return "\(x)"
}

/// Read a file's text into a project file, or explain why it is not one
/// (place three of four). Never throws: the input may be anything at all.
public func parseProjectFile(_ text: String) -> Result<ProjectFile, ProjectFileError> {
    guard let raw = JSONValue.parse(text) else { return .failure(ProjectFileError("That file is not valid JSON.")) }
    guard let o = raw.objectValue else {
        return .failure(ProjectFileError("That file is not an Atelier project file."))
    }
    guard o["kind"]?.stringValue == projectFileKind else {
        return .failure(ProjectFileError("That file is not an Atelier project file (wrong or missing kind)."))
    }
    // `typeof raw.version === 'number'` — JSON carries no NaN.
    let version = o["version"]?.finiteNumber ?? 0
    if version > Double(projectDocVersion) {
        return .failure(ProjectFileError(
            "This file was written by a newer version of Atelier (format \(projectFileNumberText(version)), "
                + "this one reads up to \(projectDocVersion)). Update the app first."
        ))
    }
    guard o["elements"]?.arrayValue != nil else {
        return .failure(ProjectFileError("The file has no overlay elements list."))
    }
    guard let rawSettings = o["settings"]?.objectValue, let aspectId = rawSettings["aspectId"]?.stringValue else {
        return .failure(ProjectFileError("The file has no project format."))
    }

    // Shape is sound; anything optional a past version did not write is
    // filled with the same default a migrated document gets. The settings
    // take the format and the shift, never the cadence.
    var timeShift = TimeShift.noShift
    if let shift = rawSettings["timeShift"]?.objectValue {
        timeShift = TimeShift(minutes: projectFileNumber(shift["minutes"]), days: projectFileNumber(shift["days"]))
    }
    var exportPrefs = ExportPrefs(fileName: nil, variants: defaultVariants())
    if let prefs = o["exportPrefs"]?.objectValue, let variants = prefs["variants"]?.arrayValue {
        exportPrefs = ExportPrefs(fileName: prefs["fileName"]?.stringValue,
                                  variants: variants.compactMap { readExportVariant($0) })
    }
    let list = readProjectElements(o["elements"])

    // Replay the document migrations: an older file must land on the current
    // shape exactly as an older stored project does.
    let shell = ProjectDoc(
        version: Int(min(max(version.rounded(.down), -1e9), 1e9)),
        id: "file", name: "", createdAt: 0, updatedAt: 0,
        settings: ProjectSettings(aspectId: aspectId, timeShift: timeShift),
        elements: list.elements,
        unreadElements: list.unread,
        guides: readGuidesState(o["guides"]),
        lutStack: readLutStack(o["lutStack"]),
        outputTransform: (o["outputTransform"]?.stringValue).flatMap(OutputTransform.init(rawValue:)) ?? .none,
        lutFilm: filmTextureOrNull(o["lutFilm"]),
        theme: readStyleTheme(o["theme"]),
        scenes: readOverlayScenes(o["scenes"]),
        outro: readOutroCard(o["outro"]),
        exportPrefs: exportPrefs
    )
    let migrated = migrateProjectDoc(shell)
    return .success(ProjectFile(
        version: projectDocVersion,
        name: o["name"]?.stringValue ?? "",
        exportedAt: o["exportedAt"]?.stringValue ?? "",
        portable: migrated.portable
    ))
}

// MARK: - place four: applying

/// Replace a document's portable half with the file's. The bound half — id,
/// source, media folder and list, trims, develops, thumbnail, creation date —
/// and the project's own NAME are kept: importing settings into "Vol du soir"
/// must not rename it after whoever exported the file.
public func applyProjectFile(_ doc: ProjectDoc, _ file: ProjectFile, now: Double = nowMillis()) -> ProjectDoc {
    var next = doc
    next.updatedAt = now
    next.settings = file.settings
    next.elements = file.elements
    next.unreadElements = file.unreadElements
    next.guides = file.guides
    next.lutStack = file.lutStack
    next.outputTransform = file.outputTransform
    next.lutFilm = file.lutFilm
    next.theme = file.theme
    // Scenes were forgotten here on the web when v12 added them, so a file
    // imported as a NEW project silently lost its intro. The outro joins both.
    next.scenes = file.scenes
    next.outro = file.outro
    next.exportPrefs = file.exportPrefs
    return next
}
